src/orion/algo/pbt/pbt.py

```diff
@@ -0,0 +1,877 @@
+        by the sequence of trial execution within a worker. This also avoids having
+        duplicate elite trials on different lineages.
+
+        Best trials may be looked for up to a ``max_depth``.
+
+        Parameters
+        ----------
+        max_depth: int or ``orion.core.worker.trial.Trial``, optional
+            The maximum depth to look for best trials. It can be an int to represent the depth
+            directly, or a trial, from which the depth will be infered. If a trial, this trial
+            should be in the Lineages. Default: None, that is, no max depth.
+        """
+        if max_depth and not isinstance(max_depth, int):
+            max_depth = self.get_lineage(max_depth).node_depth
+
+        def get_parent_at_depth(node, depth):
+            while node.node_depth > depth:
+                node = node.parent
+
+            return node
+
+        trials = []
+        for lineage in self._lineage_roots:
+            nodes = lineage.leafs
+
+            if max_depth is not None:
+                trimmed_nodes = set()
+                for node in nodes:
+                    node = get_parent_at_depth(node, max_depth)
+                    trimmed_nodes.add(node)
+                nodes = list(trimmed_nodes)
+
+            for node in nodes:
+                if node.jumps and (
+                    (max_depth is None) or (node.node_depth < max_depth)
+                ):
+                    continue
+
+                best_trial = node.get_best_trial()
+                if best_trial is not None:
+                    trials.append(best_trial)
+
+        return trials
+
+    def get_trials_at_depth(self, trial_or_depth):
+        """Returns the trials or all lineages at a given depth
+
+        Parameters
+        ----------
+        trial_or_depth: int or ``orion.core.worker.trial.Trial``
+            If an int, this represents the depth directly. If a trial, the depth will be infered
+            from it. This trial should be in the Lineages.
+
+        Raises
+        ------
+        KeyError
+            If depth is infered from trial but trial is not already registered in the Lineages
+
+        """
+        if isinstance(trial_or_depth, int):
+            depth = trial_or_depth
+        else:
+            depth = self.get_lineage(trial_or_depth).node_depth
+
+        trials = []
+        for lineage in self._lineage_roots:
+            for trial_node in lineage.get_nodes_at_depth(depth):
+                trials.append(trial_node.item)
+
+        return trials
+
+
+class Lineage(TreeNode):
+    """
+    Lineage node
+
+    The lineage node is based on :py:class:`orion.core.utils.tree.TreeNode`. It provides
+    additional methods to help represent lineages for PBT, in particular, ``fork``,
+    ``set_jump``, ``get_true_ancestor`` and ``get_best_trial``.
+
+    A lineage node can be connected to a parent and children, like a typical TreeNode, but
```

Comment:
This explain confuse me a little bit as in above `Lineages` class, we say `Each lineage is a path from a leaf trial (highest fidelity level) up to the root (lowest fidelity level)`.

Reply:
Oh right, that is very confusing. Maybe this class should be called a LineageNode?

Reply:
Sounds better :)